File: src/lib/themes.ts
import { Scene } from "./presentation";
import { callable } from "@decky/api";

/** Themes, loaded at runtime from files rather than compiled in.
 *
 * A theme used to be a TypeScript module in this bundle, which meant writing
 * one required a checkout, a toolchain and a release. Since the entire point
 * of themes is that other people write them, a theme is now a folder:
 *
 *     themes/<id>/
 *         theme.html      manifest, styles and every screen
 *         sounds/*.flac   optional
 *
 * One file plus some audio. The HTML carries its own metadata in a JSON block
 * and each screen in a <template>, so there is no second file to keep in step
 * and no build between editing a theme and seeing it.
 *
 * Loaded from ~/Documents/decky-links/themes first and the plugin's own
 * assets/themes second, so a user can shadow a shipped theme by reusing its
 * id — the cheapest possible way to tweak one.
 */

export interface ThemeSummary {
  id: string;
  name: string;
  blurb: string;
  builtin: boolean;
}

/** Per-scene settings from the manifest. Everything is optional: a theme that
 *  says nothing about a scene still gets sensible behaviour. */
export interface SceneConfig {
  enterSound?: string;
  loopSound?: string;
  /** Hold off painting until the scene has lasted this long. For scenes that
   *  can pass in a blink — a tag reads in well under 200ms, and a screen that
   *  appears for two frames reads as a fault rather than a load. */
  minVisibleMs?: number;
  fadeMs?: number;
}

export interface Theme extends ThemeSummary {
  /** Scene name to markup, from the file's <template data-scene> blocks. */
  screens: Record<string, string>;
  /** Every <style> in the file, concatenated. */
  css: string;
  /** Replacements for the plugin's own event sounds. */
  sounds: Record<string, string>;
  scenes: Record<string, SceneConfig>;
}

const listThemes = callable<[], ThemeSummary[]>("list_themes");
const readTheme = callable<[theme_id: string], { html: string } | null>("read_theme");
const readThemeAsset =
  callable<[theme_id: string, name: string], string | null>("read_theme_asset");

let summaries: ThemeSummary[] = [];
const loaded = new Map<string, Theme | null>();

/** Parse a theme file into the pieces the layer renders.
 *
 * DOMParser rather than a regex, because the markup is a whole document and
 * hand-written by people who are not us. Note that neither <script> nor an
 * inline handler in a <template> can execute: templates are inert, and React
 * sets this markup through innerHTML, which never runs script. A theme is
 * markup, and stays markup.
 */
function parse(id: string, html: string, summary: ThemeSummary): Theme {
  const doc = new DOMParser().parseFromString(html, "text/html");

  let manifest: any = {};
  const block = doc.querySelector('script[type="application/json"]#decky-theme');
  if (block?.textContent) {
    try {
      manifest = JSON.parse(block.textContent);
    } catch (e) {
      console.warn(`[ Decky Links ] Theme ${id} has an unreadable manifest:`, e);
    }
  }

  const screens: Record<string, string> = {};
  doc.querySelectorAll("template[data-scene]").forEach((el) => {
    const scene = el.getAttribute("data-scene");
    if (scene) screens[scene] = (el as HTMLTemplateElement).innerHTML;
  });

  const css = [...doc.querySelectorAll("style")].map((s) => s.textContent ?? "").join("\n");

  return {
    ...summary,
    name: manifest.name || summary.name,
    blurb: manifest.blurb || summary.blurb,
    screens,
    css,
    sounds: manifest.sounds ?? {},
    scenes: manifest.scenes ?? {},
  };
}

/** Refresh the list of installed themes. */
export async function discoverThemes(): Promise<ThemeSummary[]> {
  try {
    summaries = (await listThemes()) ?? [];
  } catch (e) {
    console.warn("[ Decky Links ] Could not list themes:", e);
    summaries = [];
  }
  return summaries;
}

export function knownThemes(): ThemeSummary[] {
  return summaries;
}

/** Load a theme's markup, once.
 *
 * Cached including the failure: a theme that will not parse must not be
 * re-fetched on every scene change, and the layer treats "no theme" the same
 * way whichever direction it arrived from.
 */
export async function loadTheme(id: string): Promise<Theme | null> {
  if (loaded.has(id)) return loaded.get(id) ?? null;

  const summary = summaries.find((t) => t.id === id)
    ?? { id, name: id, blurb: "", builtin: false };
  let theme: Theme | null = null;
  try {
    const file = await readTheme(id);
    if (file?.html) theme = parse(id, file.html, summary);
  } catch (e) {
    console.warn(`[ Decky Links ] Could not load theme ${id}:`, e);
  }
  loaded.set(id, theme);
  return theme;
}

/** Forget everything, so an edited theme shows up without a plugin restart.
 *
 * The point of a file-based format is the edit-and-look loop; a cache with no
 * way to clear it would put a restart back in the middle of that.
 */
export function forgetThemes(): void {
  loaded.clear();
}

/** The markup for a scene, or the theme's standby screen.
 *
 * The fallback is *within* the theme rather than to another one: a theme that
 * does not draw an ambient screen should idle on its own standby screen, not
 * borrow a stranger's. Falling through to Steam would be worse still — a gap
 * in a takeover is the thing this feature exists to prevent.
 */
export function screenFor(theme: Theme | null, scene: Scene): string | null {
  if (!theme) return null;
  return theme.screens[scene] ?? theme.screens[Scene.READY] ?? null;
}

export function configFor(theme: Theme | null, scene: Scene): SceneConfig {
  return theme?.scenes?.[scene] ?? {};
}

/** Fill in the handful of values a screen may ask for.
 *
 * Substitution rather than templating on purpose: a theme is markup written by
 * hand, and two placeholders someone can remember beat a syntax they have to
 * learn. Values are escaped — a game title is not markup, and "Command &
 * Conquer" should not be able to close a tag.
 */
export function fillScreen(markup: string, values: Record<string, string>): string {
  return markup.replace(/\{(\w+)\}/g, (whole, key) => {
    const value = values[key];
    if (value === undefined) return whole;
    return value
      .replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;");
  });
}

const assets = new Map<string, string>();

/** A blob URL for one of a theme's sounds.
 *
 * Themes in Documents are not served over HTTP — only the plugin's own dist/
 * is — so the bytes come over the RPC channel and become a blob here. Cached
 * by theme and name: the seek loop is fetched once, not once per disk.
 */
export async function themeAssetUrl(themeId: string, file: string): Promise<string | null> {
  const key = `${themeId}/${file}`;
  const cached = assets.get(key);
  if (cached) return cached;

  try {
    const b64 = await readThemeAsset(themeId, file);
    if (!b64) return null;
    const bytes = Uint8Array.from(atob(b64), (c) => c.charCodeAt(0));
    const url = URL.createObjectURL(new Blob([bytes]));
    assets.set(key, url);
    return url;
  } catch (e) {
    console.warn(`[ Decky Links ] Could not read ${key}:`, e);
    return null;
  }
}

File: decky_links/themes.py
"""Theme discovery: a theme is a folder, and mostly it is one file.

A theme used to be a TypeScript module compiled into the plugin's bundle,
which meant that making one required a checkout, a toolchain and a release.
The whole point of themes is that other people write them, so the format is
now data the plugin reads at runtime:

    themes/<id>/
        theme.html      the manifest, the styles and every screen
        sounds/*.flac   optional

One file plus a folder of audio. The HTML carries its own metadata in a JSON
block and each screen in a ``<template>``, so there is no second file to keep
in step with the first, and no build step between editing a theme and seeing
it. A theme with no sounds is one file.

Two directories are searched, in order:

1. ``~/Documents/decky-links/themes`` — where someone's own themes go. Same
   place the printable cards are written, because a user should have one
   folder for this plugin rather than a folder per feature.
2. The plugin's own ``assets/themes`` — what ships. Searched second so a user
   can shadow a built-in theme by using its id, which is the cheapest possible
   way to tweak one: copy it out, edit, keep the name.

Nothing here writes. Themes are read-only to the plugin, which is why none of
these functions is refused while the device is locked: reading a file the user
put there themselves gives away nothing that presenting the key would not.
"""

import base64
import json
import os
import re
from typing import Any, Dict, List, Optional

# An id is a directory name, and is used to build a path, so it is checked as
# strictly as any other untrusted string that becomes a path.
_ID = re.compile(r"^[a-z0-9][a-z0-9_-]{0,31}$")

# Asset names are equally untrusted — a theme is a folder someone downloaded.
_ASSET = re.compile(r"^[A-Za-z0-9][A-Za-z0-9._-]{0,63}$")

MANIFEST_RE = re.compile(
    r'<script[^>]+id=["\']decky-theme["\'][^>]*>(.*?)</script>',
    re.DOTALL | re.IGNORECASE,
)

# A theme is markup, not media. The cap is generous for hand-written HTML and
# nowhere near enough to be a way of loading something else through this door.
MAX_HTML_BYTES = 512 * 1024

# One sound. Big enough for a few seconds of FLAC, small enough that a theme
# cannot quietly ship a film.
MAX_ASSET_BYTES = 4 * 1024 * 1024

THEME_FILE = "theme.html"


def user_theme_dir() -> str:
    return os.path.expanduser("~/Documents/decky-links/themes")


def _plugin_theme_dirs() -> List[str]:
    """Where the shipped themes are, in both layouts they can be in.

    The packaging CLI zips a fixed allowlist that excludes a top-level
    ``assets``, so the build vendors it into ``py_modules``. A development
    checkout has it at the top level. Check both rather than guessing which
    one this is.
    """
    here = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
    return [
        os.path.join(here, "assets", "themes"),
        os.path.join(here, "py_modules", "assets", "themes"),
    ]


def search_dirs() -> List[str]:
    return [user_theme_dir(), *_plugin_theme_dirs()]


def _theme_path(theme_id: str) -> Optional[str]:
    """The directory holding this theme, or None.

    First match wins, so a user theme shadows a shipped one with the same id.
    """
    if not _ID.match(theme_id or ""):
        return None
    for root in search_dirs():
        candidate = os.path.join(root, theme_id)
        if os.path.isfile(os.path.join(candidate, THEME_FILE)):
            return candidate
    return None


def _manifest(html: str) -> Dict[str, Any]:
    """The JSON block inside a theme's HTML.

    Inside the HTML rather than beside it so a theme is one file: two files
    that must agree is two files that can disagree, and the disagreement would
    be silent.
    """
    match = MANIFEST_RE.search(html)
    if not match:
        return {}
    try:
        data = json.loads(match.group(1))
        return data if isinstance(data, dict) else {}
    except (ValueError, TypeError):
        return {}


def _read_html(path: str) -> Optional[str]:
    try:
        if os.path.getsize(path) > MAX_HTML_BYTES:
            return None
        with open(path, "r", encoding="utf-8", errors="replace") as f:
            return f.read()
    except OSError:
        return None


def list_themes() -> List[Dict[str, Any]]:
    """Every theme found, with just enough of each to draw a picker.

    Deliberately does not return the markup. The panel lists themes far more
    often than it renders one, and a list that carried every theme's whole
    body would grow with the number installed for no benefit.
    """
    found: Dict[str, Dict[str, Any]] = {}
    for root in search_dirs():
        if not os.path.isdir(root):
            continue
        for entry in sorted(os.listdir(root)):
            if entry in found or not _ID.match(entry):
                continue
            html = _read_html(os.path.join(root, entry, THEME_FILE))
            if html is None:
                continue
            manifest = _manifest(html)
            found[entry] = {
                "id": entry,
                "name": str(manifest.get("name") or entry),
                "blurb": str(manifest.get("blurb") or ""),
                # Which directory it came from, so the panel can say "yours"
                # and a user can tell a shadowing theme from the original.
                "builtin": root != user_theme_dir(),
            }
    return list(found.values())


def read_theme(theme_id: str) -> Optional[Dict[str, Any]]:
    """A theme's markup and manifest, ready to render."""
    path = _theme_path(theme_id)
    if path is None:
        return None
    html = _read_html(os.path.join(path, THEME_FILE))
    if html is None:
        return None
    return {"id": theme_id, "html": html, "manifest": _manifest(html)}


def read_asset(theme_id: str, name: str) -> Optional[str]:
    """One file from a theme's ``sounds`` folder, base64 encoded.

    Base64 over the existing RPC channel rather than a URL, because only the
    plugin's own ``dist`` is served over HTTP and a user's theme is not in it.
    The frontend turns this back into a blob and plays it.

    Both the id and the name are pattern-checked above, and the result is
    checked again with realpath: a symlink inside a downloaded theme folder is
    exactly the kind of thing that turns a read of "a sound" into a read of
    anything the plugin's user can open.
    """
    path = _theme_path(theme_id)
    if path is None or not _ASSET.match(name or ""):
        return None

    sounds = os.path.realpath(os.path.join(path, "sounds"))
    target = os.path.realpath(os.path.join(sounds, name))
    if target != sounds and not target.startswith(sounds + os.sep):
        return None
    try:
        if os.path.getsize(target) > MAX_ASSET_BYTES:
            return None
        with open(target, "rb") as f:
            return base64.b64encode(f.read()).decode("ascii")
    except OSError:
        return None
